### ISSUE TITLE
Save last run after expanding and enriching

### ISSUE THREAD

**Issue by asmodeus812** [mfussenegger/nvim-dap#1025]
### Problem Statement

It would be great if there were an option to save the last config that was run after the input from the user was collected or the config table was enriched with variables, as it is now, the last_run config that is being saved is saved before the enrichment is done, and on each successive run_last dap would keep asking for the input fields (if there are such). It might be better to store it after the enriching was done.

### Considered Alternatives

Could store the config table locally in our own config for example but i am not sure if that is robust enough.

**Comment by mfussenegger** [comment]
> It might be better to store it after the enriching was done.

It might also be worse. Depends on what the enrichment does. In some cases it needs to evaluate again to do the right thing. E.g if you're using the evaluation to compile.

Only changing the behavior is not an option. I could imagine adding a flag to re-run to choose which. But more configuration options and flags also make things more complex. So I'd like to see if there's more interest in this.

As workaround you could try something like this in your configuration:


```lua
local dap = require("dap")
local last_config = nil

---@param session Session
dap.listeners.after.event_initialized["store_config"] = function(session)
  last_config = session.config
end

function M.run_last()
  if last_config then
    dap.run(last_config)
  else
    dap.continue()
  end
end
```

I

**Comment by BickfordA** [comment]
I also ran into this trying to get `run_last` to work. I was getting errors because my config has functions for determining the path to the file being debugged and the functions couldn't be serialized when resending the config over the rpc connection:

```lua
local function json_decode(payload)
  return vim.json.decode(payload, { luanil = { object = true }})
end
local json_encode = vim.json.encode
local function send_payload(client, payload)
  local msg = rpc.msg_with_content_length(json_encode(payload))
  client.write(msg)
end
```
(`nvim-dap/lua/dap/session.lua`)

Stack trace:
```sh
        [C]: in function 'json_encode'
        .../aidan/.config/nvim/plugged/nvim-dap/lua/dap/session.lua:78: in function 'send_payload'
        .../aidan/.config/nvim/plugged/nvim-dap/lua/dap/session.lua:1694: in function 'request'
        /Users/aidan/.config/nvim/plugged/nvim-dap/lua/dap.lua:741: in function 'restart'
        /Users/aidan/.config/nvim/plugged/nvim-dap/lua/dap.lua:471: in function 'run'
        /Users/aidan/.config/nvim/plugged/nvim-dap/lua/dap.lua:526: in function 'run_last'
        /Users/aidan/.config/nvim/config/08_dap.lua:84: in function </Users/aidan/.config/nvim/config/08_dap.lua:77>
```

@mfussenegger Your snippet fixes the issue in my configuration, I can use it to cache the last run config.  If nothing else if might be nice to add this to the documentation somewhere.

Maybe there is a way we could specify which properties of the config to save (maybe in the config itself?)

**Comment by mfussenegger** [comment]
You shouldn't get errors like that using the default `dap.run_last()` as it should re-evaluate any functions before they get serialized. 

Can you share a reproduction?

**Comment by BickfordA** [comment]
This is what my config looks like. I started with the example for the lldb-vscode plugin.

```lua

local dap = require('dap')
dap.adapters.lldb = {
  type = 'executable',
  command = '/opt/homebrew/opt/llvm/bin/lldb-vscode', -- adjust as needed, must be absolute path
  name = 'lldb'
}

local dap = require('dap')

dap.configurations.cpp = {
  {
    name = 'Launch',
    type = 'lldb',
    request = 'launch',
    program = function()
      return vim.fn.input('Path to executable: ', vim.fn.getcwd() .. '/', 'file')
    end,
    cwd = '${workspaceFolder}',
    stopOnEntry = false,
    args = {},

    -- 💀
    -- if you change `runInTerminal` to true, you might need to change the yama/ptrace_scope setting:
    --
    --    echo 0 | sudo tee /proc/sys/kernel/yama/ptrace_scope
    --
    -- Otherwise you might get the following error:
    --
    --    Error on launch: Failed to attach to the target process
    --
    -- But you should be aware of the implications:
    -- https://www.kernel.org/doc/html/latest/admin-guide/LSM/Yama.html
    -- runInTerminal = false,
    initCommands = {
      'command script import /Users/aidan/ghi/lldb_formaters/py3/all.py',
      'command script import /Users/aidan/ghi/lldb_formaters/py3/ghi.py'
    }
    --initCommands = function()
    --    -- Find out where to look for the pretty printer Python module
    --    -- local rustc_sysroot = vim.fn.trim(vim.fn.system('rustc --print sysroot'))

    --    -- local script_import = 'command script import "' .. rustc_sysroot .. '/lib/rustlib/etc/lldb_lookup.py"'
    --    -- local commands_file = rustc_sysroot .. '/lib/rustlib/etc/lldb_commands'

    --    local commands = {}
    --    -- local file = io.open(commands_file, 'r')
    --    -- if file then
    --    --   for line in file:lines() do
    --    --     table.insert(commands, line)
    --    --   end
    --    --   file:close()
    --    -- end
    --    -- table.insert(commands, 1, script_import)

    --    local script_import = 'command script import /Users/aidan/ghi/lldb_formaters/py3/all.py'
    --    table.insert(commands, 1, script_import)
    --    return commands
    --  end,
      -- ...,
    }
  }

-- If you want to use this for Rust and C, add something like this:

dap.configurations.c = dap.configurations.cpp
-- dap.configurations.rust = dap.configurations.cpp

vim.keymap.set('n', '<F5>', function() require('dap').continue() end)
vim.keymap.set('n', '<F10>', function() require('dap').step_over() end)
vim.keymap.set('n', '<F11>', function() require('dap').step_into() end)
vim.keymap.set('n', '<S-F11>', function() require('dap').step_out() end)
vim.keymap.set('n', '<Leader>db', function() require('dap').toggle_breakpoint() end)
vim.keymap.set('n', '<Leader>dB', function() require('dap').set_breakpoint() end)
vim.keymap.set('n', '<Leader>lp', function() require('dap').set_breakpoint(nil, nil, vim.fn.input('Log point message: ')) end)
vim.keymap.set('n', '<Leader>dr', function() require('dap').repl.open() end)

vim.keymap.set({'n', 'v'}, '<Leader>dh', function()
  require('dap.ui.widgets').hover()
end)
vim.keymap.set({'n', 'v'}, '<Leader>dp', function()
  require('dap.ui.widgets').preview()
end)

vim.keymap.set('n', '<Leader>df', function()
  local widgets = require('dap.ui.widgets')
  widgets.centered_float(widgets.frames)
end)
vim.keymap.set('n', '<Leader>dF', function()
  local widgets = require('dap.ui.widgets')
  local my_sidebar = widgets.sidebar(widgets.frames)
  my_sidebar.open()
end)

vim.keymap.set('n', '<Leader>ds', function()
  local widgets = require('dap.ui.widgets')
  widgets.centered_float(widgets.scopes)
end)
vim.keymap.set('n', '<Leader>dS', function()
  local widgets = require('dap.ui.widgets')
  local my_sidebar = widgets.sidebar(widgets.scopes)
  my_sidebar.open()
end)

vim.keymap.set('n', '<Leader>dt', function()
  local widgets = require('dap.ui.widgets')
  widgets.centered_float(widgets.threads)
end)
vim.keymap.set('n', '<Leader>dT', function()
  local widgets = require('dap.ui.widgets')
  local my_sidebar = widgets.sidebar(widgets.threads)
  my_sidebar.open()
end)


local dap = require("dap")
local last_config = nil

---@param session Session
dap.listeners.after.event_initialized["store_config"] = function(session)
  last_config = session.config
end


vim.keymap.set('n', '<Leader>dl', function() 
  local dap = require('dap') 
  local ndap_state= dap.session()
  if ndap_state==nil then

--    if last_config then
--      dap.run(last_config)
--    else
--      dap.continue()
--    end
--
    dap.run_last()
  else 
    dap.disconnect()
    -- dap.run(last_config)
    dap.run_last()
  end
end)


```



here is a simple hello_world.cpp to debug:
```cpp
#include <iostream>

//basic hello world program
// build with:
// clang++ -g -std=c++11 -stdlib=libc++ hello_world.cpp -o hello_world
int main(){
    std::cout << "Hello World!" << std::endl;
    return 0;
}
```

I'm brand new to the plug-in, and lua in general so I could be doing something wrong. Let me know if there is more info I can provide.   You'll need clang and lldb-vscode to reproduce.

**Comment by BickfordA** [comment]
I have to run `<Leader>dl` twice, once to clear the current debugging session (paused at breakpoint), and then the second time through when restarting it throws the described stack trace.

**Comment by phcerdan** [comment]
This is a great feature, I was so annoyed having to input the arguments again and again.
I tried the `run_last` function and works perfectly.

```lua
-- Run last: https://github.com/mfussenegger/nvim-dap/issues/1025
local last_config = nil
---@param session Session
dap.listeners.after.event_initialized["store_config"] = function(session)
  last_config = session.config
end

local function debug_run_last()
  if last_config then
    dap.run(last_config)
  else
    dap.continue()
  end
end

vim.keymap.set('n', '<Leader>dl', function()
    debug_run_last()
end)

```

**Comment by mfussenegger** [comment]
@BickfordA that was a separate issue that should be fixed with https://github.com/mfussenegger/nvim-dap/issues/1101


Regarding storing the evaluated config:

I lately also had a case where it annoyed me that I needed to enter the input again and again, but I concluded that it was actually a good thing. The additional friction pushed me towards better practices. Isolating the logic I wanted to debug. I ended up creating a test case that I could then run in the debugger. Faster feedback loop and useful longterm.  

Another option is to create dedicated configuration entries in a `launch.json` if the same inputs are used over and over again.


Moral of the story: Some friction is sometimes a good thing. I won't be adding this for now. For those who really want it there's a easy enough snippet to copy.
